### PR TITLE
Add --guess-common-names

### LINKED ISSUE #46: Name-based type inference?
👋 I just found this project, and `--annotate-named-param` reminds me of a weekend research project I did for the Hypothesis ghostwriter, analyzing a few hundred million arguments in a corpus of Python code: https://github.com/HypothesisWorks/hypothesis/issues/3311.  The context is different enough between our projects that you might want to make your own decisions based on the table in that issue rather than just lifting code out of my PR (hereby offered under MIT).  

Clauses like `if name.startswith("is_") or name in BOOL_NAMES:` aren't fully expressible with `--annotate-named-param`, so this might be worth a new `--guess-common-names` argument as well as a table of suggestions, e.g. "`pattern` is usually either `str` or `re.Pattern[str]`, or sometimes `bytes`".

### PR DESCRIPTION
**Pull request by jakkdl**
Fixes #46 
leaving as draft until specs has been finalized, i.e. how to handle common names with default value of `None` and the extent of common names to use for inferring.

I decided that the least intrusive way of implementing was to hijack `annotate_optionals` [and `annotate_named_params`, so I just append the guessed names to those lists. I at first considered implementing this in `leave_Param`, and could move back there if that's preferred. It took me a while to parse `leave_Param` and I started adding comments for myself, and figured I might as well leave them.

### CODE REVIEW

**Comment by jakkdl** [comment]
Updated strategy as per @Zac-HD's suggestion and moved the logic down into `leave_Param`. Still leaving as draft for now, although it's mostly finished.
The tests could be more ambitious if you want, since this covers far from all possible branches of `guess_type_from_argname` (I guess that's what I end up doing by default when there's no coverage to keep me accountable :sweat_smile: )

File: autotyping/autotyping.py
Comment (Zac-HD) [inline autotyping/autotyping.py]: I'd just add these to float names, since typecheckers treat int as a subtype of float.

Comment (JelleZijlstra) [reply autotyping/autotyping.py]: At least for `amount` int was more common than float in your data though (which makes sense, an amount is usually going to be a count or a money amount, neither of which should be a float).

Comment (Zac-HD) [reply autotyping/autotyping.py]: If in doubt, let's drop it from the guessed names!  I'm OK with there being *some* wrong guesses, but it's worth trying to minimize them.

Comment (jakkdl) [reply autotyping/autotyping.py]: Amount being 2/3 int and 1/3 float does feel ambiguous enough that it should just be skipped. number and num are overwhelmingly int though (95% and 97% respectively) and would feel weird to give them float given the regex for `num_xxx` is `int`.
Treshold is not in the table at all, but I'll chuck it into float then.

**Comment by jakkdl** [comment]
Wrote logic for inferring simple container types, removed speculative matches, and added a lot of tests.
I also modified `_annotate_param` to handle recursive generics - so it can now do `Optional[List[int]]` etc.

File: autotyping/autotyping.py
Comment (Zac-HD) [inline autotyping/autotyping.py]:
```suggestion
    "threshold", "float",
```

(not visible: me, sweating, checking whether I typo'd this into Hypothesis.  Thankfully no)

**Review by Zac-HD — APPROVED** [review]
Tiny typo; otherwise looks good to me!

@JelleZijlstra FYI you can configure CI to need manual approval only for "new to GitHub" accounts, rather than anyone who hasn't had a PR merged previously.

File: autotyping/autotyping.py
Comment (JelleZijlstra) [inline autotyping/autotyping.py]: Please remove the fmt off, I'm a Black maintainer and I want all my code to use Black :)

File: autotyping/autotyping.py
Comment (JelleZijlstra) [reply autotyping/autotyping.py]: Maybe the typo is common enough that it should also be in the list :)

File: autotyping/autotyping.py
Comment (JelleZijlstra) [inline autotyping/autotyping.py:599]: Assuming all the containers are from `typing` works for now but isn't correct in general. For example, once we're 3.9+ we should generate `list[int]`. (And maybe we should already have an option to do that for codebases that are already 3.9+.) However, this would add some complexity. I think for now let's stick with what you have, then add a set of options like `--py39`, `--py310` to affect autotyping's output.

File: autotyping/autotyping.py
Comment (JelleZijlstra) [inline autotyping/autotyping.py]: Shouldn't this often be an Iterable or Sequence or similar? Does Zac's data cover this case?

File: autotyping/autotyping.py
Comment (JelleZijlstra) [inline autotyping/autotyping.py]: Maybe also `iter`, mapping to `Iterable`.

File: autotyping/autotyping.py
Comment (Zac-HD) [reply autotyping/autotyping.py]: I checked https://github.com/HypothesisWorks/hypothesis/issues/3311, and nope, it's not 😜

File: autotyping/autotyping.py
Comment (Zac-HD) [reply autotyping/autotyping.py]: I'm a `shed` maintainer, but still prefer to disable one-item-per-line formatting in these cases.  Your library though 🤷‍♂️

File: autotyping/autotyping.py
Comment (Zac-HD) [reply autotyping/autotyping.py:599]: I'd be inclined to just recommend `pyupgrade` to your users for that, rather than building it into `autotyping` at this stage, and then doing a hard cutover when 3.8 reaches EOL.  But ofc your call!

File: autotyping/autotyping.py
Comment (Zac-HD) [reply autotyping/autotyping.py]: Hmm, `List` is definitely the most common collection annotation, but I'm not convinced that makes it correct here.  Possibly a case where it makes sense to guess for Hypothesis, but `autotyping` would prefer something else?  If so I'd be inclined to go for either `Sequence` or not guessing at all here.

File: autotyping/autotyping.py
Comment (jakkdl) [reply autotyping/autotyping.py:599]: I also thought about this and generally agree with recommending pyupgrade. Though given that the code for it should be written for it at some point or another it doesn't really hurt (other than some code complexity) to write that now and gate the functionality behind a command line parameter.

File: autotyping/autotyping.py
Comment (jakkdl) [reply autotyping/autotyping.py]: and a better solution would be to have a spell checker running in your IDE :sweat_smile: Misspelled function parameters are awful when you try to call them and specifying keyword arguments. (I'm totally not speaking from experience as a non-native that sometimes misspells hehe).
Not fixing it also gives the author an extra chance at reading the argument and noticing the typo.

File: autotyping/autotyping.py
Comment (jakkdl) [reply autotyping/autotyping.py]: I kinda like `Sequence`, it's easy to overspecify `List` (ime) even when you don't care about modifying it and you'll get a pretty obvious error if you do modify it. But depends on how vary of false positives you want to be.

File: autotyping/autotyping.py
Comment (jakkdl) [reply autotyping/autotyping.py]: removing `fmt: off` adds 86 lines to the file currently. I could also clean up the lines slightly, they're currently somewhat hodge-podge since I wanted to initially minimize difference with the original code from hypothesis to make it easier to compare them so I added new ones on their own lines at the end - but easy to sort and squash the lines now.

Comment (jakkdl) [reply autotyping/autotyping.py]: Oh and with 114 lines of constants I'd personally either want to put them at the bottom of the file or excise them to a separate file.

**Comment by jakkdl** [comment]
Pushed a commit with minor changes from the above review

File: autotyping/autotyping.py
Comment (JelleZijlstra) [reply autotyping/autotyping.py]: A separate file for this sounds good. We're getting to 800 lines now, which is enough that splitting it up makes sense.

**Comment by jakkdl** [comment]
Moved both the constants and the function (since it as well is ~100 lines) into `autotyping.guess_type_from_argname.py`. Changed `guess_type_from_argname` to return a string instead of `NamedParam` to avoid a circular import, and did a minor restructure in `leave_Param` to avoid duplicate code calling `guess_type_from_argname`.

Also squashed the commits.

**Review by JelleZijlstra — APPROVED** [review]
Thank you!

**Pull request merged**
